.github/workflows/helm.yaml

```diff
@@ -0,0 +1,63 @@
+name: Lint and Test Charts
+
+on:
+  pull_request:
+    paths:
+      - "charts/**"
+
+jobs:
+  lint-test:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+
+      - name: Setup Helm
+        uses: azure/setup-helm@v1
+        with:
+          version: v3.5.0
+
+      - name: Setup helm-docs
+        run: |
+          HELM_DOCS_VERSION="0.15.0"
+          OS=$(uname)
+
+          curl -L -sS https://github.com/norwoodj/helm-docs/releases/download/v"${HELM_DOCS_VERSION}"/helm-docs_"${HELM_DOCS_VERSION}"_${OS}_x86_64.tar.gz \
+            | tar zx -C /tmp
+          sudo mv /tmp/helm-docs /usr/local/bin
+
+      - name: Run helm-docs
+        run: helm-docs && git diff --no-patch --exit-code
+
+      - uses: actions/setup-python@v2
+        with:
+          python-version: 3.7
+
+      - name: Set up chart-testing
+        uses: helm/chart-testing-action@v2.0.1
+
+      - name: Run chart-testing (list-changed)
+        id: list-changed
+        run: |
+          changed=$(ct list-changed --config ct.yaml)
+          if [[ -n "$changed" ]]; then
+            echo "::set-output name=changed::true"
+          fi
+
+      - name: Run chart-testing (lint)
+        run: ct lint --config ct.yaml
+
+      # TODO: If you run "helm install" in a kind, the "ct install" command will not succeed because the Pod will crash-loop.
+      #
+      # - name: Create kind cluster
+      #   uses: helm/kind-action@v1.1.0
+      #   if: steps.list-changed.outputs.changed == 'true'
+
+      # - name: Apply cert-manager CRDs
+      #   run: kubectl apply -f https://github.com/jetstack/cert-manager/releases/download/v1.3.1/cert-manager.crds.yaml
+
+      # - name: Run chart-testing (install)
+      #   run: ct install --config ct.yaml
```

Comment:
Is there any way to run it in kind?
I've temporarily commented it out because it didn't start with Chart's default value.

Reply:
I don't know what kind of problem. Could you please give me the error message?

Reply:
create kind cluster:

```console
$ kind create cluster --image kindest/node:v1.19.1
Creating cluster "kind" ...
 ✓ Ensuring node image (kindest/node:v1.19.1) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-kind"
You can now use your cluster with:

kubectl cluster-info --context kind-kind

Have a nice day! 👋
```

apply cert-manager CRDs:

```console
$ kubectl apply -f https://github.com/jetstack/cert-manager/releases/download/v1.3.1/cert-manager.crds.yaml
customresourcedefinition.apiextensions.k8s.io/certificaterequests.cert-manager.io created
customresourcedefinition.apiextensions.k8s.io/certificates.cert-manager.io created
customresourcedefinition.apiextensions.k8s.io/challenges.acme.cert-manager.io created
customresourcedefinition.apiextensions.k8s.io/clusterissuers.cert-manager.io created
customresourcedefinition.apiextensions.k8s.io/issuers.cert-manager.io created
customresourcedefinition.apiextensions.k8s.io/orders.acme.cert-manager.io created
```

helm install:

```console
$ helm install --create-namespace --namespace=topolvm-system topolvm ./charts/topolvm --set cert-manager.enabled=true
NAME: topolvm
LAST DEPLOYED: Thu Jun  3 22:24:14 2021
NAMESPACE: topolvm-system
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
# TODO
```

```console
$ kubectl get po -n topolvm-system
NAME                                              READY   STATUS             RESTARTS   AGE
topolvm-cert-manager-cainjector-97768874c-82tp9   1/1     Running            0          2m50s
topolvm-cert-manager-d746b79b-2wmpf               1/1     Running            0          2m50s
topolvm-cert-manager-webhook-566b997b7d-7zlvr     1/1     Running            0          2m50s
topolvm-controller-fb8b74bd6-25bvp                5/5     Running            0          2m50s
topolvm-controller-fb8b74bd6-hrfjs                0/5     Pending            0          2m50s
topolvm-lvmd-0-lzts7                              0/1     CrashLoopBackOff   4          2m50s
topolvm-node-tdm6h                                2/3     CrashLoopBackOff   4          2m50s
topolvm-scheduler-mhcfx                           1/1     Running            0          2m50s
```

logs:

```console
$ kubectl logs -n topolvm-system topolvm-lvmd-0-lzts7
2021-06-03T13:27:48.971167Z topolvm-lvmd-0-lzts7 lvmd info: "configuration file loaded: " device_classes="[0xc00044f7c0]" file_name="/etc/topolvm/lvmd.yaml" socket_name="/run/topolvm/lvmd.sock"
nsenter: failed to execute /sbin/lvm: No such file or directory
Error: exit status 1
exit status 1

$ kubectl logs -n topolvm-system topolvm-node-tdm6h topolvm-node
{"level":"info","ts":1622726866.7769172,"logger":"controller-runtime.metrics","msg":"metrics server is starting to listen","addr":":8080"}
{"level":"info","ts":1622726866.7775664,"logger":"controller-runtime.injectors-warning","msg":"Injectors are deprecated, and will be removed in v0.10.x"}
{"level":"info","ts":1622726866.7776146,"logger":"controller-runtime.injectors-warning","msg":"Injectors are deprecated, and will be removed in v0.10.x"}
{"level":"info","ts":1622726866.778109,"logger":"setup","msg":"starting manager"}
{"level":"info","ts":1622726866.879863,"logger":"controller-runtime.manager","msg":"starting metrics server","path":"/metrics"}
{"level":"info","ts":1622726866.879999,"logger":"controller-runtime.manager.controller.logicalvolume","msg":"Starting EventSource","reconciler group":"topolvm.cybozu.com","reconciler kind":"LogicalVolume","source":"kind source: /, Kind="}
{"level":"info","ts":1622726866.8800533,"logger":"controller-runtime.manager.controller.logicalvolume","msg":"Starting Controller","reconciler group":"topolvm.cybozu.com","reconciler kind":"LogicalVolume"}
{"level":"info","ts":1622726866.8800652,"logger":"controller-runtime.manager.controller.logicalvolume","msg":"Starting workers","reconciler group":"topolvm.cybozu.com","reconciler kind":"LogicalVolume","worker count":1}
{"level":"info","ts":1622726866.8803623,"logger":"controller-runtime.manager.controller.logicalvolume","msg":"Stopping workers","reconciler group":"topolvm.cybozu.com","reconciler kind":"LogicalVolume"}
{"level":"error","ts":1622726866.8807425,"logger":"setup","msg":"problem running manager","error":"rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial unix /run/lvmd/lvmd.sock: connect: no such file or directory\"","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\
n\t/go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132\nsigs.k8s.io/controller-runtime/pkg/log.(*DelegatingLogger).Error\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.2/pkg/log/deleg.go:144\ngithub.com/topolvm/topolvm/pkg/topolvm-node/cmd.subMain\n\t/workdir/pkg/topolvm-node/cmd/run.go:114\ngithub.com/topolvm/t
opolvm/pkg/topolvm-node/cmd.glob..func1\n\t/workdir/pkg/topolvm-node/cmd/root.go:34\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:852\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:960\ngithub.com/spf13/cobra.(*Com
mand).Execute\n\t/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:897\ngithub.com/topolvm/topolvm/pkg/topolvm-node/cmd.Execute\n\t/workdir/pkg/topolvm-node/cmd/root.go:41\nmain.main\n\t/workdir/pkg/hypertopolvm/main.go:42\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:225"}
Error: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial unix /run/lvmd/lvmd.sock: connect: no such file or directory"
rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial unix /run/lvmd/lvmd.sock: connect: no such file or directory"
```

Reply:
It is impossible to enable lvmd daemonset with kind. This is because we can't use lvm in a kind node. So we can only test lvmd runs as daemon using kind. In addition, there are some techniques when run topolvm in kind.
- lvmd is run on the host. 
  You can run it by `make start-lvmd` in e2e test.
- Configure kind to make the UNIX socket created by lvmd put on the kind node.
  https://github.com/topolvm/topolvm/blob/eccc5138e3355a832e975cb6bcdf5479cd7112bb/e2e/topolvm-cluster.yaml#L49-L50
- Configure the path for topolvm-node to communicate with lvmd.
  https://github.com/topolvm/topolvm/blob/eccc5138e3355a832e975cb6bcdf5479cd7112bb/e2e/manifests/overlays/common/node.yaml#L26-L29

Configurations for lvmd and kind are already implemented in the existing e2e, so I think it's better to reuse them. The following procedure will create an environment in which only lvmd is running.
```
cd <path to topolvm>
make -c e2e launch-kind
make -c e2e start-lvmd
make -c e2e topolvm.img
go run pkg/gencert -outdir certs
bin/kind load image-archive --name=topolvm-e2e topolvm.img
```

Reply:
To test lvmd daemonset, We use minikube. The following commands are to start up minikube and prepare the lvm.
https://github.com/topolvm/topolvm/blob/eccc5138e3355a832e975cb6bcdf5479cd7112bb/.github/workflows/main.yaml#L53-L56
There are the configurations of lvmd and topolvm-node used for testing.
https://github.com/topolvm/topolvm/blob/eccc5138e3355a832e975cb6bcdf5479cd7112bb/e2e/manifests/overlays/daemonset-lvmd/lvmd.yaml#L11-L29
https://github.com/topolvm/topolvm/blob/eccc5138e3355a832e975cb6bcdf5479cd7112bb/e2e/manifests/overlays/daemonset-lvmd/node.yaml#L1-L13

Reply:
Thanks, the test was a success! 
https://github.com/topolvm/topolvm/pull/302/commits/e90b11acac85aaca50f2db914caf245df46f7445